From dae495d7da691017b48cce017ce2c076a09b12b9 Mon Sep 17 00:00:00 2001
From: Amith A <quic_amitajit@quicinc.com>
Date: Mon, 25 Mar 2024 19:39:25 +0530
Subject: [PATCH] wifi: mac80211: Set ft_auth flag for FT roaming

A few drivers may need to know if STA is using FT auth to decide on
allowing set key.

In order to let such drivers to allow set key, add a flag 'ft_auth'
to ieee80211_sta and set it to true in case of FT roaming. This will
be used to set the need_ptk_4_way to false and thereby allowing the
driver to set key.

Signed-off-by: Amith A <quic_amitajit@quicinc.com>
---
 include/net/mac80211.h | 1 +
 net/mac80211/cfg.c     | 3 +++
 2 files changed, 4 insertions(+)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -2433,6 +2433,7 @@ struct ieee80211_link_sta {
  *	valid if the STA is a TDLS peer in the first place.
  * @mfp: indicates whether the STA uses management frame protection or not.
  * @mlo: indicates whether the STA is MLO station.
+ * @ft_auth: indicates whether the STA uses FT Authentication.
  * @max_amsdu_subframes: indicates the maximal number of MSDUs in a single
  *	A-MSDU. Taken from the Extended Capabilities element. 0 means
  *	unlimited.
@@ -2469,6 +2470,7 @@ struct ieee80211_sta {
 	bool tdls_initiator;
 	bool mfp;
 	bool mlo;
+	bool ft_auth;
 	u8 max_amsdu_subframes;
 
 	struct ieee80211_sta_aggregates *cur;
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -2453,6 +2453,9 @@ static int ieee80211_change_station(stru
 	if (err)
 		goto out_err;
 
+	if (params->sta_flags_set & BIT(NL80211_STA_FLAG_FT_AUTH))
+		sta->sta.ft_auth = true;
+
 	if (params->vlan && params->vlan != sta->sdata->dev) {
 		vlansdata = IEEE80211_DEV_TO_SUB_IF(params->vlan);
 
